cpp/src/reshape/explode.cu

```diff
@@ -36,119 +36,207 @@
-  std::unique_ptr<table> operator()(table_view const& input_table,
-                                    size_type const explode_column_idx,
-                                    bool include_pos,
-                                    rmm::cuda_stream_view stream,
-                                    rmm::mr::device_memory_resource* mr) const
-  {
-    CUDF_FAIL("Unsupported non-list column");
-
-    return std::make_unique<table>();
-  }
-};
-
-template <>
-std::unique_ptr<table> explode_functor::operator()<list_view>(
-  table_view const& input_table,
-  size_type const explode_column_idx,
-  bool include_pos,
-  rmm::cuda_stream_view stream,
-  rmm::mr::device_memory_resource* mr) const
-{
-  lists_column_view lc{input_table.column(explode_column_idx)};
-  auto sliced_child = lc.get_sliced_child(stream);
-  rmm::device_uvector<size_type> gather_map_indices(sliced_child.size(), stream);
-
-  // Sliced columns may require rebasing of the offsets.
-  auto offsets = lc.offsets_begin();
-  // offsets + 1 here to skip the 0th offset, which removes a - 1 operation later.
-  auto offsets_minus_one = thrust::make_transform_iterator(
-    offsets + 1, [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
-  auto counting_iter = thrust::make_counting_iterator(0);
-
-  rmm::device_uvector<size_type> pos(include_pos ? sliced_child.size() : 0, stream, mr);
-
-  // This looks like an off-by-one bug, but what is going on here is that we need to reduce each
-  // result from `lower_bound` by 1 to build the correct gather map. This can be accomplished by
-  // skipping the first entry and using the result of `lower_bound` directly.
-  if (include_pos) {
-    thrust::transform(
-      rmm::exec_policy(stream),
-      counting_iter,
-      counting_iter + gather_map_indices.size(),
-      gather_map_indices.begin(),
-      [position_array = pos.data(), offsets_minus_one, offsets, offset_size = lc.size()] __device__(
-        auto idx) -> size_type {
-        auto lb_idx = thrust::lower_bound(
-                        thrust::seq, offsets_minus_one, offsets_minus_one + offset_size, idx) -
-                      offsets_minus_one;
-        position_array[idx] = idx - (offsets[lb_idx] - offsets[0]);
-        return lb_idx;
-      });
-  } else {
-    thrust::lower_bound(rmm::exec_policy(stream),
-                        offsets_minus_one,
-                        offsets_minus_one + lc.size(),
-                        counting_iter,
-                        counting_iter + gather_map_indices.size(),
-                        gather_map_indices.begin());
-  }
 
+std::unique_ptr<table> build_table(table_view const& input_table,
+                                   size_type const explode_column_idx,
+                                   column_view const& sliced_child,
+                                   cudf::device_span<size_type> const gather_map,
+                                   cudf::device_span<size_type> const explode_col_gather_map,
```

Comment:
With spans, whether or not the `span` object is `const` is irrelevant. If the data pointed to by the span is meant to be immutable, then the `const` goes on the template arg.
```suggestion
                                   cudf::device_span<size_type const> gather_map,
                                   cudf::device_span<size_type const> explode_col_gather_map,
```